src/callbacks_step/amr_dg2d.jl

```diff
@@ -136,6 +137,103 @@ function refine!(u_ode::AbstractVector, adaptor, mesh::Union{TreeMesh{2}, P4estM
+            reinitialize_containers!(mesh, equations, dg, cache)
+        end
+        return
+    end
+
+    # Determine for each existing element whether it needs to be refined
+    needs_refinement = falses(nelements(dg, cache))
+    needs_refinement[elements_to_refine] .= true
+
+    # Retain current solution data
+    old_n_elements = nelements(dg, cache)
+    old_u_ode = copy(u_ode)
+    old_inverse_jacobian = copy(cache.elements.inverse_jacobian)
+    GC.@preserve old_u_ode begin # OBS! If we don't GC.@preserve old_u_ode, it might be GC'ed
+        GC.@preserve old_inverse_jacobian begin # OBS! If we don't GC.@preserve old_inverse_jacobian, it might be GC'ed
+            old_u = wrap_array(old_u_ode, mesh, equations, dg, cache)
+
+            # Loop over all elements in old container and scale the old solution by the Jacobian
+            # prior to projection
+            for old_element in 1:old_n_elements
+                for j in eachnode(dg), i in eachnode(dg)
+                    for v in eachvariable(equations)
+                        old_u[v, i, j, old_element] = (old_u[v, i, j, old_element] /
+                                                       old_inverse_jacobian[i, j,
+                                                                            old_element])
+                    end
+                end
+            end
+
+            reinitialize_containers!(mesh, equations, dg, cache)
+
+            resize!(u_ode,
+                    nvariables(equations) * nnodes(dg)^ndims(mesh) *
+                    nelements(dg, cache))
+            u = wrap_array(u_ode, mesh, equations, dg, cache)
+
+            # Loop over all elements in old container and either copy them or refine them
+            element_id = 1
+            for old_element_id in 1:old_n_elements
+                if needs_refinement[old_element_id]
+                    # Refine element and store solution directly in new data structure
+                    refine_element!(u, element_id, old_u, old_element_id,
+                                    adaptor, equations, dg)
+                    # Before `element_id` is incremented, divide by the new Jacobians on each
+                    # child element and save the result
+                    for m in 0:3 # loop over the children
+                        for j in eachnode(dg), i in eachnode(dg)
+                            for v in eachvariable(equations)
+                                u[v, i, j, element_id + m] *= (0.25 *
+                                                               cache.elements.inverse_jacobian[i,
+                                                                                               j,
+                                                                                               element_id + m])
+                            end
+                        end
+                    end
+                    element_id += 2^ndims(mesh)
```

Comment:
Maybe add a comment what is happening here (compared to line 216 ?)

Reply:
Done in 0eea3769018e18969dcff326e1152811a16e155e